sql/src/main/java/org/apache/druid/sql/calcite/planner/querygen/DruidQueryGenerator.java

```diff
@@ -0,0 +1,296 @@
+    private final RexBuilder rexBuilder;
+
+    public PDQVertexFactory(PlannerContext plannerContext, RexBuilder rexBuilder)
+    {
+      this.plannerContext = plannerContext;
+      this.rexBuilder = rexBuilder;
+    }
+
+    Vertex createVertex(PartialDruidQuery partialDruidQuery, List<Vertex> inputs)
+    {
+      return new PDQVertex(partialDruidQuery, inputs);
+    }
+
+    public class PDQVertex implements Vertex
+    {
+      final PartialDruidQuery partialDruidQuery;
+      final List<Vertex> inputs;
+
+      public PDQVertex(PartialDruidQuery partialDruidQuery, List<Vertex> inputs)
+      {
+        this.partialDruidQuery = partialDruidQuery;
+        this.inputs = inputs;
+      }
+
+      @Override
+      public DruidQuery buildQuery(boolean topLevel)
+      {
+        InputDesc input = getInput();
+        return partialDruidQuery.build(
+            input.dataSource,
+            input.rowSignature,
+            plannerContext,
+            rexBuilder,
+            !topLevel
+        );
+      }
+
+      /**
+       * Creates the {@link InputDesc} for the current {@link Vertex}.
+       */
+      private InputDesc getInput()
+      {
+        List<InputDesc> inputDescs = new ArrayList<>();
+        for (Vertex inputVertex : inputs) {
+          final InputDesc desc;
+          if (inputVertex.canUnwrapInput()) {
+            desc = inputVertex.unwrapInputDesc();
+          } else {
+            DruidQuery inputQuery = inputVertex.buildQuery(false);
+            desc = new InputDesc(new QueryDataSource(inputQuery.getQuery()), inputQuery.getOutputRowSignature());
+          }
+          inputDescs.add(desc);
+        }
+        RelNode scan = partialDruidQuery.getScan();
+        if (scan instanceof InputDescProducer) {
+          InputDescProducer inp = (InputDescProducer) scan;
+          return inp.getInputDesc(plannerContext, inputDescs);
+        }
+        if (inputs.size() == 1) {
+          return inputDescs.get(0);
+        }
+        throw DruidException.defensive("Unable to create InputDesc for Operator [%s]", scan);
+      }
+
+      /**
+       * Merges the given {@link RelNode} into the current partial query.
+       *
+       * @return the new merged vertex - or null if its not possible to merge
+       */
+      @Override
+      @Nullable
+      public Vertex mergeNode(RelNode node, boolean isRoot)
+      {
+        PartialDruidQuery newPartialQuery = mergeNode1(node, isRoot);
+        if (newPartialQuery == null) {
+          return null;
+        }
+        return createVertex(newPartialQuery, inputs);
+      }
+
+      /**
+       * Merges the given {@link RelNode} into the current {@link PartialDruidQuery}.
+       *
+       * @return the new merged {@link PartialDruidQuery} - or null if its not possible to merge.
+       */
+      @Nullable
+      public PartialDruidQuery mergeNode1(RelNode node, boolean isRoot)
+      {
+        if (accepts(node, Stage.WHERE_FILTER, Filter.class)) {
```

Comment:
In future, should we move this code to respective `DruidLogicalNode` themselves? That interface can have an `accept` method and the input to this method could be a `DruidLogicalNode` instead of a `RelNode`?

Reply:
Right now I think `PartialDruidQuery` is a way to build the `Query` - and I believe there might be other ways to build queries in the future - to keep that separation natural: I think the builder should hide away the conversion logic.

From another viewpoint: if there is another way to build `Query` (not thru `PartialDruidQuery`) ; then the nodes it uses why contain `PartialDruidQuery` related stuff?

Maybe the common denominator would be to have a target build related node classes like `PDQFilter` / `PDQProject` ... that might be another approach we could take later on....